File: programers_Lv2/괄호 회전하기.js
function solution(s) {
  let answer = 0;
  let newS = s;

  const func = (char) => {
    if (")}]".includes(char[0])) {
      return 0;
    }

    let a = (b = c = 0);
    let prev = "";
    for (let [i, v] of [...char].entries()) {
      if (prev === "[" && ")}".includes(v)) {
        answer = 0;
        return false;
      }
      if (prev === "{" && ")]".includes(v)) {
        answer = 0;
        return false;
      }
      if (prev === "(" && "}]".includes(v)) {
        return 0;
        return false;
      }

      if (v == "[") {
        a += 1;
        prev = v;
        continue;
      }
      if (v == "{") {
        b += 1;
        prev = v;
        continue;
      }
      if (v == "(") {
        c += 1;
        prev = v;
        continue;
      }

      if (v == "]" && a > 0) {
        a -= 1;
      }
      if (v == "}" && b > 0) {
        b -= 1;
      }
      if (v == ")" && c > 0) {
        c -= 1;
      }
      prev = v;
    }

    return a + b + c == 0 ? 1 : 0;
  };

  [...s].forEach((_) => {
    answer += func(newS);
    newS = newS.slice(1, s.length) + newS.slice(0, 1);
  });
  return answer;
}

console.log(solution("[](){}")); // 3
console.log(solution("}[](){")); // 3
console.log(solution("{(})")); // 0
console.log(solution("}}}")); // 0

File: programers_Lv2/괄호 회전하기.py
def solution(s):
    answer = 0
    new_s = s

    def func(t):
        if t[0] in ')}]':
            return 0
        
        a = b = c = 0
        prev = ''
        for i in range(len(t)):
            if prev == '[' and t[i] in ")}":
                return 0
            if prev == '{' and t[i] in ")]":
                return 0
            if prev == '(' and t[i] in "}]":
                return 0

            if t[i] == "[":
                a += 1
                prev = t[i]
                continue
            if t[i] == "{":
                b += 1
                prev = t[i]
                continue
            if t[i] == "(":
                c += 1
                prev = t[i]
                continue
            
            if t[i] == "]" and a > 0:
                a -= 1
            if t[i] == "}" and b > 0:
                b -= 1
            if t[i] == ")" and c > 0:
                c -= 1
            prev = t[i]

        return 1 if a + b + c == 0 else 0

    for _ in s:
        answer += func(new_s)
        new_s = new_s[1:] + new_s[:1]

    return answer


print(solution("[](){}"))  # 3
print(solution("}[](){"))  # 3
print(solution("{(})")) # 0
print(solution("}}}")) # 0
